runner/lib/helpers/buffer-line-transform.js

```diff
@@ -0,0 +1,95 @@
+/* global Buffer */
+/* eslint-disable no-underscore-dangle,no-nested-ternary,no-plusplus */
+
+import { Transform } from 'stream';
+
+export default class BufferLineTransform extends Transform {
+  /**
+   * The BufferLineTransform is reading String or Buffer content from a Readable stream
+   * and writing each line as a Buffer in object mode
+   *
+   * @param {import('./buffer-line-transform.js').BufferLineTransformOptions} [options]
+   */
+  constructor(options) {
+    const { break: breakValue, breakEncoding, ...transformOptions } =
+      options || {};
+    super({ ...transformOptions, readableObjectMode: true });
+    this._breakValue = breakValue || 10;
+    this._breakEncoding = breakEncoding;
+
+    /** @type {number} */
+    let breakLength;
+    if (!breakValue || typeof breakValue === 'number') {
+      breakLength = 1;
+    } else if (Buffer.isBuffer(breakValue)) {
+      breakLength = breakValue.length;
+    } else {
+      breakLength = Buffer.from(breakValue, breakEncoding).length;
+    }
+    this._breakLength = breakLength;
+
+    /** @type {Array<Buffer>} */
+    this._chunks = [];
+  }
+
+  /**
+   * @param {any} chunk
+   * @param {BufferEncoding | 'buffer'} encoding
+   * @param {import('stream').TransformCallback} cb
+   * */
+  _transform(chunk, encoding, cb) {
+    try {
+      /** @type {Buffer} */
+      let buf =
+        Buffer.isBuffer(chunk) || encoding === 'buffer'
+          ? chunk
+          : Buffer.from(chunk, encoding);
+
+      while (buf.length) {
+        const offset = buf.indexOf(this._breakValue, 0, this._breakEncoding);
+
+        /** @type {number} */
+        let endOffset;
+        if (offset >= 0) {
+          endOffset = offset + this._breakLength;
+          if (this._chunks.length) {
+            const concatLength = this._chunks.reduce(
+              (acc, { length }) => acc + length,
+              endOffset,
+            );
+            this._writeItem(
+              Buffer.concat(
+                [...this._chunks.splice(0, this._chunks.length), buf],
+                concatLength,
+              ),
+            );
+          } else {
+            this._writeItem(buf.subarray(0, endOffset));
+          }
+        } else {
+          endOffset = buf.length;
+          this._chunks.push(buf);
+        }
+        buf = buf.subarray(endOffset);
+      }
+      cb();
+    } catch (err) {
```

Comment:
This probably catches more error classes than intended.

Reply:
Such as? Do you see any problem with passing it all through the stream error handling logic?

Reply:
I’m sure it’s fine in its form, but:

- Node.js deopts any function that contains try/catch.
- This broad of a catch clause will also defer programming errors (e.g., ReferenceError) to the stream handler. This is approximately analogous to what happens with promises regardless, so only something to avoid if it’s easy.
- This broad of a catch obscures what would throw “invalid data type” specifically.

Reply:
Yeah this definitely assumes there are no programming errors. But I have to somehow guard against bad input, and since the whole function does input processing, I'm not sure the try/catch could be much more targeted.